### PR TITLE
add Joint_states_subscriber

### PR DESCRIPTION
**Pull request by Needrom**
for test whether it subscribe the joint_states topic

### CODE REVIEW

File: apps/joint_states_subscriber/app.c [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+	for(int i = 0; i < JOINT_DOUBLE_LEN; i++){
+		joint_states_msg.name.data[i].data = (char*) malloc(ARRAY_LEN);
+		joint_states_msg.name.data[i].size = 0;
+		joint_states_msg.name.data[i].capacity = ARRAY_LEN;
+	}
+
+//	char joint_states_name_buffer[ARRAY_LEN];
+//	joint_states_msg.name.data = joint_states_name_buffer;
+//	joint_states_msg.name.size = 7;
+//	joint_states_msg.name.capacity = ARRAY_LEN;
```
Comment (pablogs9) [inline apps/joint_states_subscriber/app.c:96]: Please remove

File: apps/joint_states_subscriber/app.c [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	RCCHECK(rclc_executor_add_subscription(&executor, &subscriber, &msg, &subscription_callback, ON_NEW_DATA));
+	RCCHECK(rclc_executor_add_subscription(&executor, &joint_states_subscriber, &joint_states_msg, &subscription_joint_state_callback, ON_NEW_DATA));
+
+	msg.data.data = (char * ) malloc(ARRAY_LEN * sizeof(char));
+	msg.data.size = 0;
+	msg.data.capacity = ARRAY_LEN;
+
+//	if(!(sensor_msgs__msg__JointState__init(&joint_states_msg))){
+//		printf("jointState message init error \r\n");		
+//	}
```
Comment (pablogs9) [inline apps/joint_states_subscriber/app.c:76]: Please remove

File: apps/joint_states_subscriber/app.c [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+	// create executor
+	rclc_executor_t executor = rclc_executor_get_zero_initialized_executor();
+	RCCHECK(rclc_executor_init(&executor, &support.context, 2, &allocator));
+	RCCHECK(rclc_executor_add_subscription(&executor, &subscriber, &msg, &subscription_callback, ON_NEW_DATA));
+	RCCHECK(rclc_executor_add_subscription(&executor, &joint_states_subscriber, &joint_states_msg, &subscription_joint_state_callback, ON_NEW_DATA));
+
+	msg.data.data = (char * ) malloc(ARRAY_LEN * sizeof(char));
+	msg.data.size = 0;
+	msg.data.capacity = ARRAY_LEN;
```
Comment (pablogs9) [inline apps/joint_states_subscriber/app.c:72]: Maybe the example is more interesting if we only subscribe to JointStates in this example. Could you refactor?

File: apps/joint_states_subscriber/app.c [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+	RCCHECK(rclc_subscription_init_default(
+		&joint_states_subscriber,
+		&node,
+		ROSIDL_GET_MSG_TYPE_SUPPORT(sensor_msgs, msg, JointState),
+		"/joint_states"));
+
+	// create executor
+	rclc_executor_t executor = rclc_executor_get_zero_initialized_executor();
+	RCCHECK(rclc_executor_init(&executor, &support.context, 2, &allocator));
+	RCCHECK(rclc_executor_add_subscription(&executor, &subscriber, &msg, &subscription_callback, ON_NEW_DATA));
```
Comment (pablogs9) [inline apps/joint_states_subscriber/app.c:67]: Maybe the example is more interesting if we only subscribe to JointStates in this example. Could you refactor?

File: apps/joint_states_subscriber/app.c [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+	// create node
+	rcl_node_t node;
+	RCCHECK(rclc_node_init_default(&node, "string_node", "", &support));
+
+	// create subscriber
+	RCCHECK(rclc_subscription_init_default(
+		&subscriber,
+		&node,
+		ROSIDL_GET_MSG_TYPE_SUPPORT(std_msgs, msg, String),
+		"/string_publisher"));
```
Comment (pablogs9) [inline apps/joint_states_subscriber/app.c:55]: Maybe the example is more interesting if we only subscribe to JointStates in this example. Could you refactor?

File: apps/joint_states_subscriber/app.c [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+rcl_subscription_t joint_states_subscriber;
+sensor_msgs__msg__JointState joint_states_msg;
+
+char test_array[ARRAY_LEN];
+
+void subscription_callback(const void * msgin)
+{
+	const std_msgs__msg__String * msg = (const std_msgs__msg__String *)msgin;
+	printf("I have heard: \"%s\"\n", msg->data.data);
+}
```
Comment (pablogs9) [inline apps/joint_states_subscriber/app.c:28]: Maybe the example is more interesting if we only subscribe to JointStates in this example. Could you refactor?

File: apps/joint_states_subscriber/app.c [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+#include <stdio.h>
+
+#define ARRAY_LEN 200
+#define JOINT_DOUBLE_LEN 20
+
+#define RCCHECK(fn) { rcl_ret_t temp_rc = fn; if((temp_rc != RCL_RET_OK)){printf("Failed status on line %d: %d. Aborting.\n",__LINE__,(int)temp_rc); return 1;}}
+#define RCSOFTCHECK(fn) { rcl_ret_t temp_rc = fn; if((temp_rc != RCL_RET_OK)){printf("Failed status on line %d: %d. Continuing.\n",__LINE__,(int)temp_rc);}}
+
+rcl_subscription_t subscriber;
+std_msgs__msg__String msg;
```
Comment (pablogs9) [inline apps/joint_states_subscriber/app.c:17]: Maybe the example is more interesting if we only subscribe to JointStates in this example. Could you refactor?

**Review by pablogs9 — CHANGES_REQUESTED** [review]
Hello, @Needrom thanks a lot! This is a super interesting example.

Maybe a couple of changes and we can merge it in the FreeRTOS apps.

Thanks a lot!

**Comment by Needrom** [comment]
already pull a new request,please check it

**Comment by pablogs9** [comment]
For the next time: if you push to the same branch the PR is updated.

**Pull request closed without merging**